File: AndroidWebBox/coconut-plugins/src/main/java/com/sniper/coconut/components/system/SystemComponent.kt
package com.sniper.coconut.components.system

import com.sniper.coconut.component.BaseComponent
import com.sniper.coconut.component.ComponentContext
import com.sniper.coconut.component.ComponentManager
import com.sniper.coconut.component.ComponentMetadata
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/**
 * System Component
 *
 * Provides version and capability introspection for the SDK.
 *
 * H5 Usage:
 *   Coconut.call('system.getVersion', {}, callback)
 *   Coconut.call('system.getComponentVersion', { name: 'network' }, callback)
 *   Coconut.call('system.getAllComponents', {}, callback)
 *   Coconut.call('system.checkCapability', { method: 'network.request' }, callback)
 */
@ComponentMetadata(
    name = "system",
    version = "1.0.0",
    description = "SDK version and capability introspection"
)
class SystemComponent : BaseComponent() {

    override val name = "system"
    override val version = "1.0.0"
    override val description = "SDK version and capability introspection"

    private var componentContext: ComponentContext? = null

    override suspend fun onInit(ctx: ComponentContext) {
        componentContext = ctx
    }

    override suspend fun handle(function: String, params: JsonObject?): JsonElement {
        return when (function) {
            "getVersion" -> getVersion()
            "getComponentVersion" -> getComponentVersion(params)
            "getAllComponents" -> getAllComponents()
            "checkCapability" -> checkCapability(params)
            else -> functionNotSupportedError(function)
        }
    }

    private fun getVersion(): JsonElement {
        val version = componentContext?.sdkVersion ?: "1.0.0"
        return buildJsonObject {
            put("sdkVersion", JsonPrimitive(version))
            put("timestamp", JsonPrimitive(System.currentTimeMillis()))
        }.let { success(it) }
    }

    private suspend fun getComponentVersion(params: JsonObject?): JsonElement {
        val name = getParam(params, "name")
        if (name.isEmpty()) {
            return error("900002", "Parameter 'name' is required")
        }

        val info = ComponentManager.getInstance().getComponentInfo(name)
            ?: return error("900001", "Component not found: $name")

        return buildJsonObject {
            put("name", JsonPrimitive(info.name))
            put("version", JsonPrimitive(info.version))
            put("description", JsonPrimitive(info.description))
            put("initialized", JsonPrimitive(info.isInitialized))
        }.let { success(it) }
    }

    private suspend fun getAllComponents(): JsonElement {
        val components = ComponentManager.getInstance().getAllComponentsInfo()
        val list = components.map { info ->
            buildJsonObject {
                put("name", JsonPrimitive(info.name))
                put("version", JsonPrimitive(info.version))
                put("description", JsonPrimitive(info.description))
                put("initialized", JsonPrimitive(info.isInitialized))
            }
        }

        return buildJsonObject {
            put("count", JsonPrimitive(list.size))
            put("components", kotlinx.serialization.json.buildJsonArray { list.forEach { add(it) } })
        }.let { success(it) }
    }

    private suspend fun checkCapability(params: JsonObject?): JsonElement {
        val method = getParam(params, "method")
        if (method.isEmpty()) {
            return error("900002", "Parameter 'method' is required")
        }

        val componentName = method.substringBefore(".")
        val component = ComponentManager.getInstance().getComponent(componentName)

        return buildJsonObject {
            put("method", JsonPrimitive(method))
            put("available", JsonPrimitive(component != null))
            put("componentRegistered", JsonPrimitive(component != null))
            put("componentInitialized", JsonPrimitive(component?.isInitialized ?: false))
        }.let { success(it) }
    }

    override suspend fun onCleanup() {
        componentContext = null
    }
}

File: iOSWebBox/CoconutSDK/Sources/CoconutSDK/Components/SystemComponent.swift
import Foundation

public class SystemComponent: BaseComponent {

    override public var name: String { "system" }
    override public var version: String { "1.0.0" }
    override public var pluginDescription: String { "SDK version and capability introspection" }

    private var componentContext: ComponentContext?

    override public func onInit(context: ComponentContext) async {
        componentContext = context
    }

    override public func handle(function: String, params: [String: Any]?) async throws -> [String: Any] {
        switch function {
        case "getVersion": return getVersion()
        case "getComponentVersion": return try getComponentVersion(params)
        case "getAllComponents": return getAllComponents()
        case "checkCapability": return try checkCapability(params)
        default: try functionNotSupportedError(function)
        }
    }

    private func getVersion() -> [String: Any] {
        let version = componentContext?.sdkVersion ?? "1.0.0"
        return success([
            "sdkVersion": version,
            "timestamp": Int(Date().timeIntervalSince1970 * 1000)
        ])
    }

    private func getComponentVersion(_ params: [String: Any]?) throws -> [String: Any] {
        let name = getParam(params, "name")
        if name.isEmpty { try error("200007", "Parameter 'name' is required") }

        guard let info = ComponentManager.shared.getComponentInfo(name: name) else {
            try error(ErrorCode.UNKNOWN_COMPONENT, "Component not found: \(name)")
        }

        return success([
            "name": info.name,
            "version": info.version,
            "description": info.description,
            "initialized": info.isInitialized
        ])
    }

    private func getAllComponents() -> [String: Any] {
        let infos = ComponentManager.shared.getAllComponentsInfo()
        let list = infos.map { info -> [String: Any] in
            return [
                "name": info.name,
                "version": info.version,
                "description": info.description,
                "initialized": info.isInitialized
            ]
        }
        return success(["count": list.count, "components": list])
    }

    private func checkCapability(_ params: [String: Any]?) throws -> [String: Any] {
        let method = getParam(params, "method")
        if method.isEmpty { try error("200007", "Parameter 'method' is required") }

        let componentName = method.components(separatedBy: ".").first ?? ""
        let component = ComponentManager.shared.getComponent(name: componentName)

        return success([
            "method": method,
            "available": component != nil,
            "componentRegistered": component != nil,
            "componentInitialized": component?.isInitialized ?? false
        ])
    }

    override public func onCleanup() async {
        componentContext = nil
    }
}
